Extending the component API

nikrowell opened the pull request:
Hi @estrattonbailey - thanks for your work on picoapp - it has become a default dependency on all of my latest WordPress projects!

I've found myself re-using a few patterns with my components and thought it might be interesting to add plugin functionality to extend / enhance the component API. Perhaps easier to explain with some usage examples:

```js
const foo = component(node, ctx, {
  findOne, 
  props,
  refs
}) => {
  const submit = findOne('button[type="submit"]')
  const { name, email } = refs
  console.log(`start preloading ${props.images.length} images...`)
})

const app = picoapp({foo})

// adds findOne and findAll utility functions that query in context of the current node
app.use(node => ({
  findOne: s => node.querySelector(s),
  findAll: s => Array.from(node.querySelectorAll(s))
})

// pass WordPress / ACF fields to the component via data-props attribute
// data-props='<?php echo json_encode(['images' => $images]) ?>'
app.use(node => {
  const props = JSON.parse(node.dataset.props || '{}')
  return {props}
})

// pull out all child elements with a ref attribute 
// (inspired by https://github.com/terkelg/facon)
app.use(node => {
  const refs = Array.from(node.querySelectorAll('[ref]')).reduce((acc, child) => {
    const key = child.getAttribute('ref').trim()
    acc[key] = acc[key]
      ? Array.isArray(acc[key])
        ? [...acc[key], child]
        : [acc[key], child]
      : child;
    return acc
  }, {})
  return {refs}
})

app.mount()
```

I've added a new test case, but probably could be more thorough. I'm also passing the extensions / plugin api through as `args` - not crazy about that name, but I'd likely be destructuring this in my component definition anyway. Curious what you think - and no worries if it doesn't fit your vision of **picoapp** (or there are issues with this implementation that I'm not seeing 😬). 

Cheers!

nikrowell commented:
@estrattonbailey Interested in adding something like this to picoapp or should I close?

estrattonbailey commented:
Yo! Sorry for the delay. I moved from NYC to Chicago in early December, then holidays... I'm sure you know how it goes lol.

This is neat! Love the `facon` usage, I love that lib too. I've got quite a bit of things to get to in the next week or so, so will get back to you. At first glance, it looks great :)

nikrowell commented:
Thinking through this some more, perhaps it make more sense to merge objects returned from "plugins" into the `ctx` that gets passed to the component? 

One situation that tripped me up was using the ctx event bus to trigger a modal. Passing modal options through `emit()` merged those in as global state, and persisted on the next call - expected behavior after realizing what was going on, but using plugins, you could replace on / off / emit with something like [mitt](https://github.com/developit/mitt), if desired.

nikrowell commented:
Thanks for the feedback and great suggestions! Good point on the `findOne` utility. Will work through these changes in the next week or two (friends coming to visit, so TBD on timing).

I like the simplicity of an optional third param to `emit(name, state, merge)`. In my case, I had modal options like `closeOnEscape` that, if I passed on the first call and omitted on the second, the second modal would receive the merged / preserved state from the first call.

nikrowell commented:
I think I've got this all buttoned up! I updated plugin functions to receive both `node` and `ctx`, merged plugin results / extensions into `ctx`, added a test assertion to ensure context internals aren't overridden and updated README docs to include a section on using Plugins. Let me know if you see anything else you'd like changed!

estrattonbailey commented:
@nikrowell just published 3.5.0!
